### DISCUSSION TITLE
Monitoring of MQTT clients at the level of "classes of clients" (based on client ID regex patterns)

### DISCUSSION THREAD

**Discussion by chrihartl** [vernemq/vernemq#2316] [Ideas]
### Environment
Any, I believe.

### Expected behaviour
VerneMQ allows  monitoring at the level of individual clients (if there is just a handful) or at the level of "classes of clients", based on client ID patterns (if there are  large numbers of clients).

### Actual behaviour
The "expected" (desired) behavior (feature) is currently not implemented.

### Idea
In order to make individual VerneMQ clients (client IDs) monitorable through Prometheus but still avoid the problem of monitoring the status of an unbounded number of clients, provide the following feature:

- By default, monitor each client individually (see below).
- Optionally allow configuring a list of "monitoring client id patterns" (a list of regex expressions against which client ids are matched), such that client IDs are grouped instead of considered individually for the purpose of monitoring.

For each monitored individual client - or, if configured, class of clients - you would get various metrics, for example
CLIENT_STATUS{client_id_pattern='a', status='connected'}  (counter)
CLIENT_CONNECTION_TIME{client_id_pattern='a'} (histogram)

### Rationale
If there is just a handful (bounded number) of clients, the default monitoring behaviour would just work fine.

On the other hand, for the case of a large-scale system one would configure something like:
```
monitoring_client_id_patterns:
  - 'africa-*'
  - 'arab-states-*'
  - 'asia-pacific-*'
  - 'europe-*'
  - 'middle-east-*'
  - 'north-america-*'
  - 'south-latin-america-*'
```

and get the above Prometheus metrics (in the form of labelled counter metrics or histograms) for client classes that potentially wrap large numbers of clients.

This could be a pretty useful extension of the already awesome set of VerneMQ metrics available through Prometheus.

**Comment by ioolkos** [comment]
@chrihartl thank you for your suggestion and taking the time to write it up!

My first questions would be around scalability (as you're aware) and what data points to collect. Beyond that, I see an undetermined amount of experiments and work to make this into a useful feature. Internally, Verne would probably have to do a lot of additional accounting (depending on desired data points), implementing it's own plugin hooks to collect events.

I cannot do much more than acknowledge this as a feature request. No work will happen on it from my side right now.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages (.deb/.rpm/Docker) requires a subscription.

**Comment by chrihartl** [comment]
@ioolkos appreciate your answer, thanks!

Regarding the data points to collect, it could be enough to just track the number of connected clients per defined client group.
This would only hit the CPU whenever a client connects or disconnects.

Something along these Python lines of code:

```python

from prometheus_client import Gauge

# The configured list of client group names (e.g. given as client id regex patterns)
client_group_names = ['asia', 'europe']  # Actually, take this from the VerneMQ config.

# A function that returns True if a given client id belongs to a defined group
def client_group_match(client_id: str, group_name: str):
    # To be done.
    raise NotImplementedError

# A labelled gauge that tracks, for each client group, the number of connected clients.
clients = Gauge(
    name='clients',
    documentation=' VerneMQ Clients',
    labelnames=('client_group_name', )
)

# A function that gets called whenever a client CONNECTS.
def client_connect_callback(client_id: str):
    # For each client group match of this client id, we INCREMENT the corresponding Prometheus gauge value.
    for group_name in client_group_names:
        if client_group_match(client_id, group_name):
            group_clients: Gauge = clients.labels(client_group_name=group_name)
            group_clients.inc()

# A function that gets called whenever a client DISCONNECTS.
def client_disconnect_callback(client_id: str):
    # For each client group match of this client id, we DECREMENT the corresponding Prometheus gauge value.
    for group_name in client_group_names:
        if client_group_match(client_id, group_name):
            group_clients: Gauge = clients.labels(client_group_name=group_name)
            group_clients.dec()
```
